pkg/render/manager_route_config.go

```diff
@@ -0,0 +1,416 @@
+		volumeName := fmt.Sprintf("cm-%s", name)
+
+		// Use a different folder for secrets and configmaps to avoid mounting collisions.
+		mountLocation := fmt.Sprintf("%s/%s", configMapFolder, name)
+		volumeMount := corev1.VolumeMount{Name: volumeName, MountPath: mountLocation, ReadOnly: true}
+
+		volume := corev1.Volume{
+			Name: volumeName,
+			VolumeSource: corev1.VolumeSource{
+				ConfigMap: &corev1.ConfigMapVolumeSource{
+					LocalObjectReference: corev1.LocalObjectReference{Name: name},
+					DefaultMode:          &defaultMode,
+				},
+			},
+		}
+
+		builder.volumeMounts = append(builder.volumeMounts, volumeMount)
+		builder.volumes = append(builder.volumes, volume)
+
+		// Use mapToSortedArray to ensure the ordering is the same and the annotation value isn't changed because of
+		// map ordering differences.
+		builder.hashableContent = append(builder.hashableContent, mapToSortedArray(configMap.Data))
+		builder.mountedConfigMaps[name] = struct{}{}
+	}
+
+	return fmt.Sprintf("%s/%s/%s", configMapFolder, name, key), nil
+}
+
+func (builder *voltronRouteConfigBuilder) mountSecretReference(name, key string) (string, error) {
+	defaultMode := int32(420)
+
+	secret := builder.secrets[name]
+
+	if secret == nil {
+		return "", fmt.Errorf("the contents for Secret '%s' wasn't provided, and is needed to generate annotations", name)
+	}
+
+	if _, ok := builder.mountedSecrets[name]; !ok {
+		// Prefix the volume name to avoid collisions with secrets that have the same name.
+		volumeName := fmt.Sprintf("scrt-%s", name)
+
+		// Use a different folder for secrets and configmaps to avoid mounting collisions.
+		mountLocation := fmt.Sprintf("%s/%s", secretsFolder, name)
+		volumeMount := corev1.VolumeMount{Name: volumeName, MountPath: mountLocation, ReadOnly: true}
+
+		volume := corev1.Volume{
+			Name: volumeName,
+			VolumeSource: corev1.VolumeSource{
+				Secret: &corev1.SecretVolumeSource{
+					SecretName:  name,
+					DefaultMode: &defaultMode,
+				},
+			},
+		}
+
+		builder.volumeMounts = append(builder.volumeMounts, volumeMount)
+		builder.volumes = append(builder.volumes, volume)
+
+		// Use mapToSortedArray to ensure the ordering is the same and the annotation value isn't changed because of
+		// map ordering differences.
+		builder.hashableContent = append(builder.hashableContent, mapToSortedArray(secret.Data))
+		builder.mountedSecrets[name] = struct{}{}
+	}
+
+	return fmt.Sprintf("%s/%s/%s", configMapFolder, name, key), nil
+}
+
+// mapToSortedArray returns an array with the contents of the map, ordered lexicographically by the map keys.
+// This provides a consistent ordering to the contents of the map.
+func mapToSortedArray[E any](m map[string]E) []E {
+	keys := make([]string, len(m))
+	i := 0
+	for key := range m {
+		keys[i] = key
+		i++
+	}
+
+	sort.Strings(keys)
+
+	i = 0
+	values := make([]E, len(keys))
+	for _, key := range keys {
+		values[i] = m[key]
+		i++
+	}
+
+	return values
```

Comment:
FWIW, I think this is also not a completely correct way to generate a hash for the annotations. We're ignoring the keys when building the annotation, so the following result in the same hash:

```yaml
aa: bb
mm: cc
zz: dd
```
```yaml
bb: bb
nn: cc
rr: dd
```

The returned hashable content should include the keys in some capacity as well.

Reply:
That is a good point.

Reply:
The key is now part of the annotation name.